quickwit-indexing/src/source/kafka_source.rs

```diff
@@ -0,0 +1,959 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use crate::models::RawDocBatch;
+use crate::source::IndexerMessage;
+use crate::source::Source;
+use crate::source::SourceContext;
+use crate::source::TypedSourceFactory;
+use anyhow::{bail, Context};
+use async_trait::async_trait;
+use futures::StreamExt;
+use itertools::Itertools;
+use quickwit_actors::ActorExitStatus;
+use quickwit_actors::Mailbox;
+use quickwit_metastore::checkpoint::{Checkpoint, CheckpointDelta, PartitionId, Position};
+use rdkafka::config::{ClientConfig, RDKafkaLogLevel};
+use rdkafka::consumer::stream_consumer::StreamConsumer;
+use rdkafka::consumer::{Consumer, ConsumerContext, Rebalance};
+use rdkafka::error::{KafkaError, KafkaResult};
+use rdkafka::message::BorrowedMessage;
+use rdkafka::topic_partition_list::TopicPartitionList;
+use rdkafka::util::Timeout;
+use rdkafka::{ClientContext, Message, Offset};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use std::collections::HashMap;
+use std::fmt;
+use std::time::Duration;
+use tracing::{debug, info, warn};
+
+/// Required parameters for instantiating a `KafkaSource`.
+#[derive(Clone, Deserialize, Serialize)]
+pub struct KafkaSourceParams {
+    bootstrap_servers: String,
+    group_id: String,
+    topic: String,
+    enable_partition_eof: Option<bool>,
+}
+
+impl KafkaSourceParams {
+    pub fn new(
+        bootstrap_servers: String,
+        group_id: String,
+        topic: String,
+        enable_partition_eof: Option<bool>,
+    ) -> Self {
+        KafkaSourceParams {
+            bootstrap_servers,
+            group_id,
+            topic,
+            enable_partition_eof,
+        }
+    }
+}
+
+/// Factory for instantiating a `KafkaSource`.
+pub struct KafkaSourceFactory;
+
+#[async_trait]
+impl TypedSourceFactory for KafkaSourceFactory {
+    type Source = KafkaSource;
+    type Params = KafkaSourceParams;
+
+    async fn typed_create_source(
+        params: KafkaSourceParams,
+        checkpoint: Checkpoint,
+    ) -> anyhow::Result<Self::Source> {
+        KafkaSource::try_new(params, checkpoint).await
+    }
+}
+
+struct KafkaSourceContext;
+
+impl ClientContext for KafkaSourceContext {}
+
+impl ConsumerContext for KafkaSourceContext {
+    fn pre_rebalance(&self, rebalance: &Rebalance) {
+        info!("Pre rebalance {:?}", rebalance);
+    }
+
+    fn post_rebalance(&self, rebalance: &Rebalance) {
+        info!("Post rebalance {:?}", rebalance);
+    }
+
+    fn commit_callback(&self, result: KafkaResult<()>, _offsets: &TopicPartitionList) {
+        info!("Committing offsets: {:?}", result);
```

Comment:
No, we don't commit since we rely entirely on checkpoints for committing. I was mostly interested in the callbacks for acting upon rebalancing. The method is `&self`, though.